rust/phoenix-channel/src/login_url.rs

```diff
@@ -0,0 +1,224 @@
+use base64::{engine::general_purpose::STANDARD, Engine};
+use secrecy::{CloneableSecret, ExposeSecret as _, SecretString, Zeroize};
+use sha2::Digest as _;
+use std::net::{Ipv4Addr, Ipv6Addr};
+use url::Url;
+use uuid::Uuid;
+
+// From https://man7.org/linux/man-pages/man2/gethostname.2.html
+// SUSv2 guarantees that "Host names are limited to 255 bytes".
+// POSIX.1 guarantees that "Host names (not including the
+// terminating null byte) are limited to HOST_NAME_MAX bytes".  On
+// Linux, HOST_NAME_MAX is defined with the value 64, which has been
+// the limit since Linux 1.0 (earlier kernels imposed a limit of 8
+// bytes)
+//
+// We are counting the nul-byte
+#[cfg(not(target_os = "windows"))]
+const HOST_NAME_MAX: usize = 256;
+
+#[derive(Clone)]
+pub struct LoginUrl {
+    url: Url,
+    host: String,
+}
+
+impl Zeroize for LoginUrl {
+    fn zeroize(&mut self) {
+        let placeholder = Url::parse("http://a.com").expect("placeholder URL to be valid");
+        let _ = std::mem::replace(&mut self.url, placeholder);
+    }
+}
+
+impl CloneableSecret for LoginUrl {}
+
+impl LoginUrl {
+    pub fn client<E>(
+        url: impl TryInto<Url, Error = E>,
+        firezone_token: &SecretString,
+        device_id: String,
+        device_name: Option<String>,
+        public_key: [u8; 32],
+    ) -> std::result::Result<Self, LoginUrlError<E>> {
+        let external_id = hex::encode(sha2::Sha256::digest(device_id));
```

Comment:
No action needed - In some parts of the client this was changing to "firezone-id" but I actually think "device_id" is more intuitive since it is device-specific and it's not related to the actor or anything.

I don't remember why we were changing it to "firezone-id"